# Python Cheatsheet 101

This jupyter notebook covers the most basic principles in Python and presents them in a very simple, yet effective manner! I tried my best to cover the basic pillars of Pyton by including the following topics:

1. Print Statments, it's parameters & Comments
    * Print function
    * `Sep` & `End` parmeters
    * Comments
2. Data types and Variables
    * Integer, Floats, Strings and Booleans
    * Single Variable & multiple assignment
3. Arithmetic and Operators
    * Arithmetic & Math functions
    * Operators (>, <, <=, >=)
3. Conditionals
    * If Statements
    * Switch Case Statements
4. Loops
    * While Loop
    * For Loop
    * Nested Loop
5. Strings
    * Multple ways to assign them
    * String slicing
    * String methods
6. Lists
    * Creating lists
    * List slicing and iterating
    * List methods
    * Nested lists
7. Tuples
    * Creating tuples
    * Slicing tuples (Similar to Strings & Lists)
    * Tuple methods
8. Dictionaries
    * Creating dictionaries
    * Dictionary slicing and iterating
    * Dictionary methods
9. Sets
    * Creating sets
    * Set methods
10. Functions
    * Creating a function, parameters and return
    * `*args` & `**kwargs`
    * Lambda Function
13. Comprehensions
    * List Comprehensions
    * Dictionary Comprehensions


Well, this is it. I won't be covering Object Oriented Concepts as it is a lengthy topic. Will do in another dedicated cheatsheeet.

## 1. Print Statments, it's Parameters & Comments

In [144]:
# Print Statement
print("This is a Print Statement!")

This is a Print Statement!


In [145]:
# The `sep` parameter
print('python','cool') # Default `sep` value is set to a space (i.e " ")
print('python','cool.com', sep='@')
print('1','1', '2023', sep='/')

# Given 2 strings and a given sep (separator) it joins them with the given sep.

python cool
python@cool.com
1/1/2023


In [146]:
# The `end` parameter
print("Python", end='@') 
print("CheatSheet")

# Ends the output with '@', however, default value is 

Python@CheatSheet


In [147]:
# PS. this stuff after the "#" is called a comment.

## 2. Data types and Variables

#### The classical datatypes

In [148]:
x = str("Hello World")
print(type(x))

x = int(20) 
print(type(x))

x = float(20.5) 
print(type(x))

x = bool(5) 
print(type(x))

<class 'str'>
<class 'int'>
<class 'float'>
<class 'bool'>


#### Data structural datatypes

In [149]:
x = list(("apple", "banana", "cherry")) 
print(type(x))

x = tuple(("apple", "banana", "cherry")) 
print(type(x))

x = range(6)
print(type(x))

x = dict(name="John", age=36)
print(type(x))

x = set(("apple", "banana", "cherry"))
print(type(x))

x = frozenset(("apple", "banana", "cherry"))  
print(type(x))

<class 'list'>
<class 'tuple'>
<class 'range'>
<class 'dict'>
<class 'set'>
<class 'frozenset'>


#### Some other datatypes (Binary)

In [150]:
x = complex(1j) 
print(type(x))

x = bytes(5)
print(type(x))

x = bytearray(5)
print(type(x))

x = memoryview(bytes(5))
print(type(x))

<class 'complex'>
<class 'bytes'>
<class 'bytearray'>
<class 'memoryview'>


#### Variables

In [151]:
name = "Sally"
print(name)

Sally


In [152]:
age = 23
print(age)

23


Variables are used to store data, kinda like a container that holds up water or spices in your home!

## 3. Arithmetic and Operators
1. Arithmetic
2. Math module
3. Operators

#### 1. Arithmetic

In [153]:
# Addition
print(5 + 3)

8


In [154]:
# Subtraction
print(11-5)

6


In [155]:
# Multiplication
print(2 * 10)

20


In [156]:
# Division
print(20 / 5)

4.0


In [157]:
# Floor Division (Divides 2 numbers, but answer is always in integer (whole number) form)
print(20 // 6)

3


In [158]:
# Modulus (What is the REMAINDER?)
print(5 % 5) # 0

print(5 % 3) # 1

0
2


In [159]:
# Exponentiation (Sqaure)
print(2 ** 3) # 2*2*2 (2^3)

8


#### Math module

In [160]:
# Import math Library
import math

In [161]:
math.e

2.718281828459045

In [162]:
math.pi

3.141592653589793

In [163]:
math.tau 

6.283185307179586

In [164]:
math.inf

inf

You can read more about the `math` module here: https://docs.python.org/3/library/math.html

#### Operators

In [165]:
print(18 > 2) # Greater than

True


In [166]:
print(18 < 2) # Less than

False


In [167]:
print(18 == 18) # Equal to

True


In [168]:
print(18 >= 18) # Greater than or Equal to

True


In [169]:
print(19 <= 18) # Less than or Equal to

False


## 4. Conditionals
1. If Statements
2. Shorthand
3. Case/Switch Statements

### If Statements

In [170]:
age = 80

In [171]:
if age >= 18 and age <= 60:
    print("You're an adult")
elif age >= 60:
    print("You're old!")
elif age >= 13:
    print("Congrats! You're a teen!")
else:
    print("Still a child my freind.")

You're old!


The above example uses an if-statement & also makes use of the `and` operator. Other operators you can use are `or` and `not`.

In [172]:
# Shorthand
a = 60
b = 60

print("A is LARGEST") if a > b else print("NET 0") if a == b else print("B is LARGEST")

NET 0


### Switch Case
Now let's have a look at a new addition to Python - **`Switch Case Statemets`**, along awaited request from the Python community and in Python **V.3.10** we get this!

In [173]:
lang = "Python"

match lang:
    case "JavaScript":
        print("You can become a web developer.")

    case "Python":
        print("You can become a Data Scientist")

    case "PHP":
        print("You can become a backend developer")
    
    case "Solidity":
        print("You can become a Blockchain developer")

    case "Java":
        print("You can become a mobile app developer")
    case _:
        print("The language doesn't matter, what matters is solving problems.")

You can become a Data Scientist


A pretty understandable syntax here. We start the case statement with the **match** keyword, followed by the name of the variable. Then you use the **case** keyword as shown.

When ending the case statement, you can use the **underscore (_)** and close off as shown, however this is optional just like the **else** keyword in the **if-else** statements.

## 5. Loops
1. While Loop
2. For Loop
3. Nested Loop

### While Loop
First and foremost, we have the while loop - used when we have to iterate until a conditon is fullfilled. Beow is an example.

In [174]:
count = 0
while count < 5:
    print(count)
    count += 1

0
1
2
3
4


This loop will print the numbers `0` through `4`. The loop will continue as long as the value of count is less than 5. The line `count += 1` is shorthand for `count = count + 1` and is used to increment the value of count by 1 after each iteration of the loop.

In [175]:
# A more complex example - finding factorial
n = 5
result = 1
while n > 1:
    result *= n
    n -= 1
    
print(result)

120


This example defines a factorial program that takes an integer n as input and returns the factorial of n. The factorial of a number is the product of all the integers from 1 to that number. For example, the factorial of 5 is 5 * 4 * 3 * 2 * 1 = 120.

The `while` loop in this example continues as long as `n` is greater than 1. The value of `result` is updated on each iteration of the loop by multiplying it by the current value of `n`, and then `n` is decremented by 1. When the loop exits, `result` will contain the factorial of the input `n`.

### For Loop
`for` loops are used when you have a block of code that you want to execute a specific number of times. They are useful for iterating over a sequence of elements, such as a list or a range of numbers.

For example, you might use a for loop to iterate over a list of names and print each name to the console, or to iterate over a range of numbers and calculate the sum of the numbers.

`for` loops are also useful for performing an action on each element of a sequence, such as modifying the element or adding it to a new list.

Overall, `for` loops are an efficient and concise way to perform a repetitive task in your code.

In [176]:
for i in range(5):
    print(i)

0
1
2
3
4


This loop will print the numbers 0 through 4. The `range(5)` function returns a sequence of numbers from 0 to 4, and the for loop iterates over this sequence, assigning each number to the loop variable `i` in turn. On each iteration of the loop, the value of `i` is printed.

In [177]:
for i in range(5):
    print("Python IS AMAZING!")

Python IS AMAZING!
Python IS AMAZING!
Python IS AMAZING!
Python IS AMAZING!
Python IS AMAZING!


Instead of printing the `i` variable, I just printed a string 5 times, by writing it once only!

#### Range?
The `range()` function in Python returns a sequence of numbers, starting from 0 by default, and increments by 1 (also by default), and ends at a specified number - 1 / (n - 1).

For example, `range(5)` returns a sequence of numbers from 0 to 4, `range(2, 6)` returns a sequence of numbers from 2 to 5, and `range(3, 10, 2)` returns a sequence of numbers from 3 to 9, incrementing by 2 on each iteration.

`range()` is commonly used in for loops to specify the number of iterations, as shown above.

In [178]:
# A slightly complex example - counting vowels

vowels = "aeiouAEIOU" # We'll use this string for comparison
string = "Python is Cool" # We'll be iterating over this string

vowel_count = 0 # The `i` variable in this case

for char in string:
    if char in vowels:
        vowel_count += 1

print(vowel_count)

4


This example defines that a `string` is taken as input and returns the number of vowels in the string. The `for` loop iterates over **each character in the string**, and the `if statement` inside the loop checks if the character is a vowel. If it is, the `vowel count` is incremented by `1`. When the loop exits, the vowel count is printed.

This for loop demonstrates how to iterate over the characters in a string and perform an action on each character. It also shows how to use an if statement inside a for loop to selectively perform an action based on the value of the loop variable.

### Nested Loops

A nested loop is a loop inside another loop. In other words, a loop is "nested" within another loop.

Nested loops are used when you want to perform a task multiple times with different combinations of input values. For example, you might use a nested loop to iterate over the rows and columns of a two-dimensional data structure, such as a matrix.

In [179]:
number_of_tables = 2
muliplicands = 16 

for i in range(1, number_of_tables):
    for j in range(0, muliplicands + 1): # I added one so the loop would go till the exact desired number & not n-1
        product = i * j
        print(f"{i} x {j} = {product}")

1 x 0 = 0
1 x 1 = 1
1 x 2 = 2
1 x 3 = 3
1 x 4 = 4
1 x 5 = 5
1 x 6 = 6
1 x 7 = 7
1 x 8 = 8
1 x 9 = 9
1 x 10 = 10
1 x 11 = 11
1 x 12 = 12
1 x 13 = 13
1 x 14 = 14
1 x 15 = 15
1 x 16 = 16


The outer for loop iterates over the numbers 1 through 10, and the inner for loop iterates over the numbers 1 through 10 for each value of the outer loop.

This example shows how a nested loop can be used to generate a simple, yet useful output by iterating over a range of numbers and performing a calculation on each combination of values. It also demonstrates how the `f-strings` feature of Python can be used to easily format the output.

## 6. Strings

1. Multple ways to assign them
2. String slicing
3. String methods


### Creatint Strings
There are several ways to assign a string in Python. The most common way is to use single or double quotes:

In [180]:
string_1 = 'Hello, world!'
string_2 = "Hello, world!"

You can also use triple quotes to create a multi-line string:

In [181]:
string_3 = """Hello,
world!"""
string_3

'Hello,\nworld!'

You can use the `+` operator to concatenate (join) two or more strings:

In [182]:
string_4 = "Hello, " + "world!"
string_4

'Hello, world!'

You can also use the `*` operator to repeat a string a certain number of times:

In [183]:
string_5 = "Hello " * 3
string_5

'Hello Hello Hello '

This will create a string that contains `"Hello, "` repeated `3` times.

Finally, you can use the `str()` function to convert a non-string value (such as a number) to a string:

In [184]:
number = 42
string_6 = str(number)

print(string_6)
type(string_6)

42


str

### Slicing Strings

In Python, you can use slicing to extract a part of a string (a substring). Slicing uses the syntax `string[start:stop:step]`, where start is the index of the first character to include in the substring, stop is the index of the first character to exclude from the substring, and step is the number of characters to skip between consecutive characters in the substring.

Here is an example of slicing a string in Python:

In [185]:
string = "Python is Snakey"
substring = string[0:6] # (n - 1), index starts from 0
print(substring)  # prints "Python"

Python


Below are some more examples of string slicing.

In [186]:
string = "Hello, world!"

# Extract the first 5 characters
substring = string[:5]
print(substring)  # prints "Hello"

Hello


In [187]:
# Extract the last 5 characters
substring = string[-5:]
print(substring) 

orld!


In [188]:
# Extract the character at index 5
substring = string[5]
print(substring)

,


In [189]:
# Extract the characters at even indices (0, 2, 4, etc.)
substring = string[::2]
print(substring) 

Hlo ol!


In [190]:
# Extract the characters at odd indices (1, 3, 5, etc.)
substring = string[1::2]
print(substring)

el,wrd


In [191]:
# Extract the characters in reverse order
substring = string[::-1]
print(substring)

!dlrow ,olleH


In [192]:
# Finding an element within a string
string = "Python is beautiful"
print("Happy") if "Python" in string else print("Sad")

string = "Java is beautiful"
print("Happy") if "Python" in string else print("Sad")

Happy
Sad


These examples demonstrate how you can use slicing to extract different parts of a string, using various combinations of start, stop, and step indices. Slicing is a powerful and convenient way to manipulate strings in Python.

### String Methods

In [193]:
string = "Hello, world!"

In [194]:
# Convert the string to uppercase
uppercase_string = string.upper()
print(uppercase_string)

HELLO, WORLD!


In [195]:
# Convert the string to lowercase
lowercase_string = string.lower()
print(lowercase_string)

hello, world!


In [196]:
# Check if the string starts with a specific substring
result = string.startswith("Hello")
print(result)

True


In [197]:
# Check if the string ends with a specific substring
result = string.endswith("world!")
print(result)

True


In [198]:
# Find the index of the first occurrence of a specific substring
index = string.index("world")
print(index)

7


In [199]:
# Replace a specific substring with another string
new_string = string.replace("world", "universe")
print(new_string)

Hello, universe!


In [200]:
# Check if a string has a digit
result = any(char.isdigit() for char in string)
print(result)

string = "Hello World/2"
result = any(char.isdigit() for char in string)
print(result)

False
True


## 7. Lists
1. Creating lists
2. List slicing and iterating
3. List methods
4. Nested lists

In Python, a list is an ordered collection of items. Lists can contain elements of different data types, including integers, floating-point numbers, strings, and even other lists. Lists are created using square brackets (`[]`) to enclose a comma-separated sequence of values.

Lists are an important data type in Python and are used to store and manipulate collections of values. Lists are dynamic, which means you can add or remove elements from a list after it is created. You can also access, modify, and iterate over the elements of a list using indexing, slicing, and looping.

In [201]:
# Create a list of integers
list_1 = [1, 2, 3, 4, 5]

# Create a list of strings
list_2 = ['apple', 'banana', 'cherry']

# Create a list of mixed data types
list_3 = [1, 'apple', 3.14, True]

You can also use the `list()` function to create a list from an iterable, such as a string or another list. For example:

In [202]:
# Create a list from a string
string = "hello"
list_4 = list(string)
print(list_4)

['h', 'e', 'l', 'l', 'o']


In [203]:
# Create a list from another list
test_lst = [4, 5, "Lists"]
list_5 = list(test_lst)
print(list_5)

[4, 5, 'Lists']


### List Slicing
In Python, you can use slicing to extract a part of a list (a sublist). Slicing uses the syntax `list[start:stop:step]`, where start is the index of the first element to include in the sublist, stop is the index of the first element to exclude from the sublist, and step is the number of elements to skip between consecutive elements in the sublist.

***Same as we did in Strings.***

Here are some examples of slicing lists in Python:

In [204]:
# Create a list of integers
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [205]:
# Extract the first 5 elements
sublist = numbers[:5]
print(sublist)

[1, 2, 3, 4, 5]


In [206]:
# Extract the last 5 elements
sublist = numbers[5:]
print(sublist)

[6, 7, 8, 9, 10]


In [207]:
# Extract the elements at even indices (0, 2, 4, etc.)
sublist = numbers[::2]
print(sublist)

[1, 3, 5, 7, 9]


In [208]:
# Extract the elements at odd indices (1, 3, 5, etc.)
sublist = numbers[1::2]
print(sublist) 

[2, 4, 6, 8, 10]


In [209]:
# Extract the elements in reverse order
sublist = numbers[::-1]
print(sublist)

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]


These examples demonstrate how you can use slicing to extract different parts of a list, using various combinations of start, stop, and step indices. Slicing is a powerful and convenient way to manipulate lists in Python.

### List Iterating
We can print out the elements in a list via a `for` loop.

In [210]:
fruits = ["Apple", "Kiwi", "Orange"]
for fruit in fruits:
    print(fruit)

Apple
Kiwi
Orange


In [211]:
# Something more fun - Use a for loop to print the sum of all elements
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
sum = 0
for number in numbers:
    sum += number
print(sum) 

55


### List Methods

In [212]:
# Create a list of integers
numbers = [1, 2, 3, 4, 5]

In [213]:
# Append an element to the end of the list
numbers.append(6)
print(numbers) 

[1, 2, 3, 4, 5, 6]


In [214]:
# Insert an element at a specific index
numbers.insert(2, 7)
print(numbers) 

[1, 2, 7, 3, 4, 5, 6]


In [215]:
# Remove an element by value
numbers.remove(3)
print(numbers)

[1, 2, 7, 4, 5, 6]


In [216]:
# Remove an element by index
numbers.pop(4)
print(numbers) 

[1, 2, 7, 4, 6]


In [217]:
# Find the index of an element
index = numbers.index(7)
print(index)

2


In [218]:
# Count the number of occurrences of an element
count = numbers.count(7)
print(count)

1


In [219]:
# Sort the list in ascending order
numbers.sort()
print(numbers)

[1, 2, 4, 6, 7]


In [220]:
# Reverse the order of the list
numbers.reverse()
print(numbers)

[7, 6, 4, 2, 1]


These examples demonstrate how you can use various list methods to manipulate and modify a list in Python. There are many other list methods available, and you can find a complete list in the Python documentation.

### Nested Lists
In the example below, the nested list contains three inner lists, each with three elements. You can access an element of the inner list using two sets of square brackets and the indices of the element.

Nested lists can be useful for organizing and storing data in a hierarchical or multi-dimensional structure. You can use loops and list indexing to iterate over and manipulate the elements of a nested list.

In [221]:
# Create a list of lists
nested_list = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]

# Access an element of the inner list
print(nested_list[1][1])

5


## 8. Tuples

1. Creating tuples
2. Slicing tuples (Similar to Strings & Lists)
3. Tuple methods


In Python, a tuple is an `immutable` sequence data type. Tuples are similar to lists, but they are created using parentheses (`()`) instead of square brackets (`[]`). Tuples are ordered collections of elements, and they can contain elements of different data types, like strings, integers, and floating-point numbers.

### Creating Tuples

In [222]:
# Create a tuple of integers
tuple1 = (1, 2, 3, 4, 5)

# Create a tuple of strings
tuple2 = ('apple', 'banana', 'cherry')

# Create a tuple of mixed data types
tuple3 = (1, 'apple', 3.14, True)

# Create an empty tuple
tuple4 = ()

Tuples are immutable, which means you cannot add or remove elements from a tuple, or modify the elements of a tuple. However, you can access, slice, and iterate over the elements of a tuple using indexing, slicing, and looping. You can also use the len() function to determine the length of a tuple, and the tuple() function to convert an iterable, such as a list, into a tuple.

Tuples are often used to store fixed sets of data, such as the names of the days of the week or the coordinates of a point in space. They are also used to return multiple values from a function and to unpack multiple values into separate variables.

### Slicing Tuples

Same as Lists and Strings. Won't repeat.

### Tuple Methods
Tuples are immutable, which means they do not have any methods that modify the tuple itself. However, tuples do have several built-in functions that you can use to perform various operations on tuples.

Here are some examples of tuple functions in Python:

In [223]:
# Create a tuple of integers
tuple_1 = (1, 2, 3, 4, 5)

In [224]:
# Use the len() function to determine the length of the tuple
length = len(tuple_1)
print(length)

5


In [225]:
# Use the min() and max() functions to find the minimum and maximum elements
minimum = min(tuple_1)
maximum = max(tuple_1)

print(minimum)
print(maximum)

1
5


In [226]:
# Use the sorted() function to create a new sorted tuple
sorted_tuple = tuple(sorted(tuple_1))
print(sorted_tuple)

(1, 2, 3, 4, 5)


In [227]:
# Use the tuple() function to create a tuple from a list
list_1 = [1, 2, 3]
tuple_2 = tuple(list_1)
print(tuple_2)

(1, 2, 3)


## 9. Dictionaries

1. Creating dictionaries
2. Iterating over dictionaries
3. Dictionary methods

*In Python, you cannot slice a dictionary because dictionaries are unordered collections of key-value pairs and do not have a defined sequence. Slicing requires a defined sequence of elements, such as a list or a tuple, but dictionaries do not have an inherent order to their elements.*


In Python, a dictionary is an unordered collection of data values that are stored as key-value pairs. Dictionaries are created using curly braces (`{}`) and are accessed using keys, which are unique identifiers for each value in the dictionary. Dictionaries can contain keys and values of different data types, like strings, integers, and floating-point numbers.

### Creating Dictionaries

In [228]:
# Create a dictionary with string keys and integer values
dict1 = {'a': 1, 'b': 2, 'c': 3}

# Create a dictionary with integer keys and string values
dict2 = {1: 'apple', 2: 'banana', 3: 'cherry'}

# Create a dictionary with mixed data types
dict3 = {'a': 1, 2: 'apple', 'c': 3.14, 4: True}

# Create an empty dictionary
dict4 = {}


In a dictionary, each key is associated with a single value, and you can use the keys to access the values in the dictionary. You can use the len() function to determine the number of key-value pairs in a dictionary, and the del statement to delete a key-value pair from a dictionary. You can also use loops and dictionary methods to iterate over and manipulate the elements of a dictionary.

Dictionaries are useful for storing data that needs to be quickly accessed using keys, such as the results of a database query or the contents of a configuration file. They are also used to count the frequency of elements in a list or to create a mapping between two sets of data.

In [229]:
# Create a dictionary with string keys and integer values
dict1 = {'a': 1, 'b': 2, 'c': 3}

# Use a for loop to print the keys and values of the dictionary
for key, value in dict1.items():
    print(f'{key}: {value}')

a: 1
b: 2
c: 3


This code uses a for loop and the `items()` method to iterate over the key-value pairs of the dictionary and print the `keys` and `values`. You can use similar techniques to perform various operations on the elements of a dictionary, such as filtering or transforming the values or creating a new dictionary based on the keys or values.

### Dictionary Methods

In [230]:
# Create a dictionary with string keys and integer values
dict1 = {'a': 1, 'b': 2, 'c': 3}

In [231]:
# Use the keys() method to get a view of the keys
keys = dict1.keys()
print(keys)

dict_keys(['a', 'b', 'c'])


In [232]:
# Use the values() method to get a view of the values
values = dict1.values()
print(values)

dict_values([1, 2, 3])


In [233]:
# Use the items() method to get a view of the key-value pairs
items = dict1.items()
print(items)

dict_items([('a', 1), ('b', 2), ('c', 3)])


In [234]:
# Use the get() method to retrieve a value for a key
value = dict1.get('a')
print(value)

1


In [235]:
# Use the get() method with a default value
value = dict1.get('d', 0)
print(value)

0


In [236]:
# Use the pop() method to remove a key-value pair
value = dict1.pop('a')
print(value)
print(dict1) 

1
{'b': 2, 'c': 3}


In [237]:
# Use the update() method to add or update key-value pairs
dict1.update({'b': 3, 'd': 4})
print(dict1)

{'b': 3, 'c': 3, 'd': 4}


In [238]:
# Use the clear() method to remove all key-value pairs
dict1.clear()
print(dict1) 

{}


These examples demonstrate how you can use dictionary methods to retrieve, modify, and manipulate the elements of a dictionary in Python. You can find a complete list of dictionary methods in the Python documentation.

## 10. Sets
* Creating sets
* Set Methods (Union, Intersection, etc.)

In Python, a set is an unordered collection of unique elements. Sets are created using curly braces (`{}`) or the `set()` function, and they can contain elements of different data types, like strings, integers, and floating-point numbers.

### Creating Sets

In [240]:
# Create a set using curly braces
set1 = {1, 2, 3, 4, 5}

# Create a set using the set() function
set2 = set([1, 2, 3, 4, 5])

# Create a set from a string
set3 = set('abcdef')

# Create an empty set
set4 = set()

Sets do not allow duplicate elements, and they do not preserve the order of the elements. You can use the `len()` function to determine the number of elements in a set, and the in operator to check if an element is in a set. You can also use `loops` and `set methods` to iterate over and manipulate the elements of a set.

Sets are useful for storing data that needs to be unique, such as the names of students in a class or the words in a document. They are also used to perform set operations, such as `intersection`, `union`, and `difference`.

### Iterating over Sets

In [247]:
# Create a set of integers
set_1 = {1, 2, 3, 4, 5}

# Use a for loop to iterate over the elements of the set
for element in set_1:
    print(element)

1
2
3
4
5


### Set Methods

In [248]:
# Create two sets of integers
set_1 = {1, 2, 3, 4, 5}
set_2 = {3, 4, 5, 6, 7}

In [249]:
# Use the union() method to create a new set with the elements from both sets
set_3 = set1.union(set_2)
print(set_3)

{1, 2, 3, 4, 5, 6, 7}


In [250]:
# Use the intersection() method to create a new set with the elements that are in both sets
set_4 = set_1.intersection(set_2)
print(set_4) 

{3, 4, 5}


In [251]:
# Use the difference() method to create a new set with the elements that are in set1 but not set2
set_5 = set1.difference(set_2)
print(set_5)

{1, 2}


In [252]:
# Use the symmetric_difference() method to create a new set with the elements that are in either set1 or set2, but not both
set_6 = set_1.symmetric_difference(set_2)
print(set_6)

{1, 2, 6, 7}


In [253]:
# Use the add() method to add an element to the set
set_1.add(6)
print(set_1)

{1, 2, 3, 4, 5, 6}


In [254]:
# Use the remove() method to remove an element from the set
set_1.remove(6)
print(set_1)

{1, 2, 3, 4, 5}


In [255]:
# Use the pop() method to remove an element from the set and return it
element = set_1.pop()
print(element)  # prints 1
print(set_1)

1
{2, 3, 4, 5}


In [256]:
# Use the clear() method to remove all elements from the set
set_1.clear()
print(set_1) 

set()


These examples demonstrate how you can use set methods to modify and manipulate the elements of a set in Python. You can find a complete list of set methods in the Python documentation.

## 11. Function
1. Creating a function, parameters and return, and **doc** strings
2. `*args` & `**kwargs`
3. Lambda Function

In Python, a function is a block of code that performs a specific task and can be reused multiple times in a program. Functions help you to organize and structure your code, making it easier to read and maintain.

### Creating Functions

In [257]:
def greet(name):
    """This function greets the user by name."""
    print(f'Hello, {name}!')

# Call the function with a string argument
greet('Alice')
greet('Bob')

Hello, Alice!
Hello, Bob!


This code defines a function called `greet()` that takes a single argument name and prints a greeting message to the user. The function definition consists of the `def` keyword, the `function name`, and a `pair of parentheses` that may contain `zero or more arguments`. The function body is indented and contains the statements that are executed when the function is called.

The `greet()` function can be called with different arguments to greet different users. You can also use default arguments to specify a default value for an argument, and keyword arguments to specify arguments by name.

Functions can return a value to the caller using the `return` statement. For example:

In [258]:
def add(x, y):
    """This function adds two numbers and returns the result."""
    result = x + y
    return result

# Call the function with two arguments and store the result
z = add(3, 4)
print(z)

7


This code defines a function called add() that takes two arguments x and y and returns the sum of these numbers to the caller. The function is called with two arguments and the result is stored in a variable z, which is then printed to the console.

Functions are a useful way to organize and reuse code in Python.

### DOC STRINGS
**Question**: What is the string just below it's decleration in triple quotes?

A docstring (short for documentation string) is a string literal that appears as the first statement in a Python function, method, class, or module. Docstrings are used to document the purpose and behavior of Python code.

Docstrings are optional in Python, but they are a good practice to include in your code because they provide documentation for users of your code. Docstrings can be used to explain the purpose, arguments, and return value of a function, and to provide examples of usage. They can also be used to document the attributes and methods of a class, and the variables and functions of a module.

### `*Args` and `**Kwargs`

In Python, the `*args` and `**kwargs` syntax allows you to pass a variable number of arguments to a function.

`*args` is used to pass a non-keyworded, variable-length argument list to a function. It is used to pass a list of arguments to a function as a tuple.

Here is an example of using `*args` in a function definition:

In [261]:
def foo(*args):
    """This function prints the arguments passed to it."""
    for arg in args:
        print(arg)

# Call the function with multiple arguments
foo(1, 2, 3, 4, 5)

1
2
3
4
5


In this example, the `foo()` function takes a variable number of arguments and prints them to the console. The `*args` syntax in the function definition indicates that the function can take any number of arguments, which are then passed to the function as a tuple.

`**kwargs` is used to pass a keyworded, variable-length argument list to a function. It is used to pass a dictionary of arguments to a function.

Here is an example of using `**kwargs` in a function definition:

In [262]:
def func(**kwargs):
    """This function prints the keyword arguments passed to it."""
    for key, value in kwargs.items():
        print(f'{key}: {value}')

# Call the function with multiple keyword arguments
func(x = 1, y = 2, z = 3)

x: 1
y: 2
z: 3


In this example, the `foo()` function takes a variable number of keyword arguments and prints them to the console. The `**kwargs` syntax in the function definition indicates that the function can take any number of keyword arguments, which are then passed to the function as a dictionary.

You can use the `*args` and `**kwargs` syntax together in the same function definition, like this:

In [263]:
def func(x, *args, **kwargs):
    """This function prints the arguments and keyword arguments passed to it."""
    print(f'x: {x}')
    for arg in args:
        print(arg)
    for key, value in kwargs.items():
        print(f'{key}: {value}')

# Call the function with multiple arguments and keyword arguments
func(1, 2, 3, 4, 5, y = 6, z = 7)

x: 1
2
3
4
5
y: 6
z: 7


In this example, the `foo()` function takes a required argument `x`, followed by a variable number of non-keyworded arguments and a variable number of keyword arguments. You can use the `*args` and `**kwargs` syntax to define flexible and reusable functions that can accept a variable number of arguments and keyword arguments in Python.

### Lambda Functions
In Python, a lambda function is a small anonymous function that is defined without a name using the `lambda` keyword. Lambda functions are used to perform simple operations, and they are often used as anonymous functions (functions without a name) or as arguments to higher-order functions (functions that accept other functions as arguments).

Lambda functions are often used in combination with higher-order functions such as `map()`, `filter()`, and `reduce()`, which operate on sequences of data. 

Here is an example of using a lambda function with the `map()` function:


In [266]:
# Define a lambda function that takes a number and returns its square
# Define a lambda function that takes a number and returns its square
square = lambda x: x**2

# Create a list of numbers
numbers = [1, 2, 3, 4, 5]

# Use the map() function to apply the lambda function to each element of the list
squares = map(square, numbers)
print(list(squares))

[1, 4, 9, 16, 25]


In this example, the `square()` function is defined as a lambda function that takes a single argument `x` and returns its square. The `map(`) function applies the lambda function to each element of the numbers list and returns a new iterable object with the squared values. The `list()` function is used to create a list from the iterable object.

Lambda functions are useful when you need to define a simple function for a short period of time and do not want to define a full function using the `def` keyword. They can also be used to avoid defining and naming unnecessary functions, and to improve the readability and clarity of your code.

## 12. Comprehensions
1. List Comprehensions
2. Dictionary Comprehensions

### List Comprehension
List comprehensions are a concise way to create lists in Python. They are a convenient and efficient way to transform one list into another, or to create a new list based on a condition or a transformation.

Here are some examples of list comprehensions in Python:

In [267]:
# Create a new list with the squares of the numbers from 1 to 10
squares = [x**2 for x in range(1, 11)]
print(squares)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


In [268]:
# Create a new list with the even numbers from 1 to 10
evens = [x for x in range(1, 11) if x % 2 == 0]
print(evens)

[2, 4, 6, 8, 10]


In [269]:
# Create a new list with the first names of a list of persons
persons = [('Alice', 'Smith'), ('Bob', 'Jones'), ('Charlie', 'Brown')]
first_names = [name[0] for name in persons]
print(first_names) 

['Alice', 'Bob', 'Charlie']


### Dictionary Comprehensions
Dictionary comprehensions are a concise way to create dictionaries in Python. They are similar to list comprehensions, but they create dictionaries instead of lists, and they use curly braces `{}` instead of square brackets `[]`.

Here are some examples of dictionary comprehensions in Python:

In [272]:
# Create a new dictionary with the squares of the numbers from 1 to 10 as keys and their square roots as values
squares = {x: x**3 for x in range(1, 11)}
print(squares)

{1: 1, 2: 8, 3: 27, 4: 64, 5: 125, 6: 216, 7: 343, 8: 512, 9: 729, 10: 1000}


In [276]:
# Create a new dictionary with the lowercase letters of the English alphabet as keys and their ASCII codes as values
letters = {chr(x): x for x in range(ord('a'), ord('z')+1)}
print(letters)

{'a': 97, 'b': 98, 'c': 99, 'd': 100, 'e': 101, 'f': 102, 'g': 103, 'h': 104, 'i': 105, 'j': 106, 'k': 107, 'l': 108, 'm': 109, 'n': 110, 'o': 111, 'p': 112, 'q': 113, 'r': 114, 's': 115, 't': 116, 'u': 117, 'v': 118, 'w': 119, 'x': 120, 'y': 121, 'z': 122}


In [277]:
# Create a new dictionary with the last names of a list of persons as keys and their first names as values
persons = [('Alice', 'Smith'), ('Bob', 'Jones'), ('Charlie', 'Brown')]
names = {name[1]: name[0] for name in persons}
print(names)  # prints

{'Smith': 'Alice', 'Jones': 'Bob', 'Brown': 'Charlie'}
